pkg/services/ngalert/state/state_tracker.go

```diff
@@ -0,0 +1,90 @@
+package state
+
+import (
+	"fmt"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/services/ngalert/eval"
+	"github.com/grafana/grafana/pkg/services/ngalert/models"
+	"sync"
+)
+
+type AlertState struct {
+	UID     string
+	CacheId string
+	Labels  data.Labels
+	State   eval.State
+	Results []eval.State
+}
+
+type cache struct {
+	cacheMap map[string]AlertState
+	mu       sync.Mutex
+}
+
+var stateCache cache
+
+func Init() {
```

Comment:
nit: You can probably turn this function into a constructor by using a:

```

type StateTracker struct {
	stateCache cache
}

func NewStateTracker() *StateTracker {
	return &StateTracker{
		stateCache: cache,
	}
}
```

Then, within the NG service (Grafana uses a service-based architecture with inject as a DI framework so you get lots of benefits when you make your struct part of the main service):

```
type AlertNG struct {
	Cfg             *setting.Cfg                            `inject:""`
	DatasourceCache datasources.CacheService                `inject:""`
	RouteRegister   routing.RouteRegister                   `inject:""`
	SQLStore        *sqlstore.SQLStore                      `inject:""`
	DataService     *tsdb.Service                           `inject:""`
	Alertmanager    *notifier.Alertmanager                  `inject:""`
	DataProxy       *datasourceproxy.DatasourceProxyService `inject:""`
	Log             log.Logger
	schedule        schedule.ScheduleService
	stateTracker *state.StateTracker
}
```

and finally 
```
// Init initializes the AlertingService.
func (ng *AlertNG) Init() error {
	ng.Log = log.New("ngalert")
   ng.stateTracker = state.NewStateTracker()
	baseInterval := baseIntervalSeconds * time.Second
	....
}
```

Reply:
Done